google-cloud-firestore/src/main/java/com/google/cloud/firestore/BasePath.java

```diff
@@ -123,16 +127,42 @@ boolean isPrefixOf(BasePath<B> path) {
   public int compareTo(@Nonnull B other) {
     List<String> thisSegments = this.getSegments();
     List<String> otherSegments = other.getSegments();
+
     int length = Math.min(thisSegments.size(), otherSegments.size());
     for (int i = 0; i < length; i++) {
-      int cmp = thisSegments.get(i).compareTo(otherSegments.get(i));
+      int cmp = compareSegments(thisSegments.get(i), otherSegments.get(i));
       if (cmp != 0) {
         return cmp;
       }
     }
     return Integer.compare(thisSegments.size(), otherSegments.size());
   }
 
+  private int compareSegments(String lhs, String rhs) {
+    boolean isLhsNumeric = isNumericId(lhs);
+    boolean isRhsNumeric = isNumericId(rhs);
+
+    if (isLhsNumeric && !isRhsNumeric) { // Only lhs is numeric
+      return -1;
+    } else if (!isLhsNumeric && isRhsNumeric) { // Only rhs is numeric
+      return 1;
+    } else if (isLhsNumeric && isRhsNumeric) { // both numeric
+      return Long.compare(extractNumericId(lhs), extractNumericId(rhs));
+    } else { // both non-numeric
+      return lhs.compareTo(rhs);
+    }
+  }
+
+  // Checks if a segment is a numeric ID (starts with "__id" and ends with "__").
```

Comment:
Nit: Is there a reason not to write these method comment as a javadoc? `/** ... */`?